openedx_external_enrollments/tests/api/v0/test_views.py

```diff
@@ -0,0 +1,689 @@
+
+        with LogCapture(level=logging.INFO) as log_capture:
+            response = self.base._post_enrollment(data, course_settings)
+            log4 = 'External enrollment response for [{}] -- {}'.format(self.base.__str__(), data)
+            log_capture.check(
+                ('openedx_external_enrollments.api.v0.views', 'INFO', log1),
+                ('openedx_external_enrollments.api.v0.views', 'INFO', log2),
+                ('openedx_external_enrollments.api.v0.views', 'INFO', log3),
+                ('openedx_external_enrollments.api.v0.views', 'INFO', log4),
+            )
+
+        self.assertEqual(response, (data, status.HTTP_200_OK))
+        url_mock.assert_called_with(course_settings)
+        data_mock.assert_called_with(data, course_settings)
+        post_mock.assert_called_with(
+            url=url,
+            headers=headers,
+            json_data=data,
+        )
+
+        log_details = {
+            'request_payload': data,
+            'url': url,
+            'course_advanced_settings': course_settings,
+            'response': data,
+        }
+
+        request_log = EnrollmentRequestLog.objects.filter(request_type=str(self.base), details=log_details)
+        self.assertEqual(len(request_log), 1)
+
+        headers_mock.side_effect = NotImplementedError('My test error')
+
+        with LogCapture(level=logging.INFO) as log_capture:
+            response = self.base._post_enrollment(data, course_settings)
+            log4 = 'Failed to complete enrollment. Reason: {}'.format('My test error')
+            log_capture.check(
+                ('openedx_external_enrollments.api.v0.views', 'INFO', log1),
+                ('openedx_external_enrollments.api.v0.views', 'INFO', log2),
+                ('openedx_external_enrollments.api.v0.views', 'INFO', log3),
+                ('openedx_external_enrollments.api.v0.views', 'ERROR', log4),
+            )
+
+        log_details['response'] = {'error': log4}
+        request_log = EnrollmentRequestLog.objects.filter(request_type=str(self.base), details=log_details)
+        self.assertEqual(len(request_log), 1)
+
+    def test_get_enrollment_data(self):
+        """Testing _get_enrollment_data method."""
+        self.assertRaises(NotImplementedError, self.base._get_enrollment_data)
+
+    def test_get_enrollment_headers(self):
+        """Testing _get_enrollment_headers method."""
+        self.assertRaises(NotImplementedError, self.base._get_enrollment_headers)
+
+    def test_get_enrollment_url(self):
+        """Testing _get_enrollment_url method."""
+        self.assertRaises(NotImplementedError, self.base._get_enrollment_url)
+
+
+class EdxEnterpriseExternalEnrollmentTest(TestCase):
+    """Test class for EdxEnterpriseExternalEnrollment class."""
+
+    def setUp(self):
+        """Set test database."""
+        self.base = EdxEnterpriseExternalEnrollment()
+
+    def test_get_enrollment_data(self):
```

Comment:
what do you mean with invalid values ?  if `course_settings` is not a dict or if the key doesn't exist

Reply:
Keys are not called as required. You are just testing the "happy path" but what about the path when you have invalid data, Should it be a failure? Should it be continue? @andrey-canon

Reply:
could you write an example?, because I don't understand what you suggest

Reply:
```
course_settings = {
    'fake_key_name': 'test_course_run_id',
    'External_enrollment_mode_override': 'test_mode',
}
```
What I mean is, what should happen when you don't have the data you expect? In theory, should it continue? or what you think should happen. @andrey-canon If you think it should raise an exception, we need to write a test case that prove it.

Reply:
The method doesn't verify the keys, since the course_settings can contain a lot information, if data and course_setting are dict the method won't raise an exception due to it makes a secure verification, and if the key doesn't exist the value will be None

```
expected_data = {
            'course_run_id': None,
            'course_mode': None,
            'user_email': None,
            'is_active': None,
        }
```

so do you suggest to test this case ??

Reply:
Ok, but what should happens when values are None? again, should it fail? should enrollment be executed? just to add a "no-happy-path" test. @andrey-canon

Reply:
looks like you are talking about integration test, this is a unit test for the method, if the enrollment must be executed with None values depends on other method, the construction of this method is not fail when the values are None,if we want to do that we have to change the method and that is not part of this pr

Reply:
No, it's just simply, make a test to check how _get_enrollment_data should behave when you pass wrong or incorrect data, that's all. Again not just the happy-path. @andrey-canon

Reply:
the behavior is the same, I don't understand what you want. This is the call method  `self.base._get_enrollment_data(data, course_settings)` what are the values for data and course-settings that are not the 'happy-path' ??? we have already discussed some of that possible values, but I don't see another possible options.